zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -133,17 +182,55 @@ impl Chain {
 
     /// Fork a chain at the block with the given hash, if it is part of this
     /// chain.
-    pub fn fork(&self, fork_tip: block::Hash) -> Result<Option<Self>, ValidateContextError> {
+    ///
+    /// The note commitment trees must be the trees of the finalized tip.
+    pub fn fork(
+        &self,
+        fork_tip: block::Hash,
+        sprout_note_commitment_tree: sprout::tree::NoteCommitmentTree,
+        sapling_note_commitment_tree: sapling::tree::NoteCommitmentTree,
+        orchard_note_commitment_tree: orchard::tree::NoteCommitmentTree,
+    ) -> Result<Option<Self>, ValidateContextError> {
         if !self.height_by_hash.contains_key(&fork_tip) {
             return Ok(None);
         }
 
-        let mut forked = self.clone();
+        let mut forked = self.with_trees(
+            sprout_note_commitment_tree,
+            sapling_note_commitment_tree,
+            orchard_note_commitment_tree,
+        );
 
         while forked.non_finalized_tip_hash() != fork_tip {
             forked.pop_tip();
         }
 
+        // Rebuild the note commitment trees, starting from the finalized tip tree.
+        // TODO: change to a more efficient approach by removing nodes
+        // from the tree of the original chain (in `pop_tip()`).
+        // See https://github.com/ZcashFoundation/zebra/issues/2378
+        for block in forked.blocks.values() {
+            for transaction in block.block.transactions.iter() {
+                for sprout_note_commitment in transaction.sprout_note_commitments() {
+                    forked
+                        .sprout_note_commitment_tree
+                        .append(sprout_note_commitment);
+                }
```

Comment:
Similar to previous comments, let's avoid adding sprout until it actually works.